core/clingutils/src/TClingUtils.cxx

```diff
@@ -4587,8 +4587,16 @@ clang::QualType ROOT::TMetaUtils::ReSubstTemplateArg(clang::QualType input, cons
          }
       } else if (auto const declguide = llvm::dyn_cast<clang::CXXDeductionGuideDecl>(replacedDeclCtxt)) {
          replacedCtxt = llvm::dyn_cast<clang::ClassTemplateDecl>(declguide->getDeducedTemplate());
+      } else if (auto const ctdecl = llvm::dyn_cast<clang::ClassTemplateDecl>(replacedDeclCtxt)) {
+         replacedCtxt = ctdecl;
       } else {
-         replacedCtxt = llvm::dyn_cast<clang::ClassTemplateDecl>(replacedDeclCtxt);
+         std::string astDump;
+         llvm::raw_string_ostream ostream(astDump);
+         instance->dump(ostream);
+         ostream.flush();
+         ROOT::TMetaUtils::Warning("ReSubstTemplateArg","Unexpected type of declaration context for template parameter: %s.\n\tThe responsible class is:\n\t%s\n",
+                                   replacedDeclCtxt->getDeclKindName(), astDump.c_str());
+         replacedCtxt = nullptr;
       }
```

Comment:
Hi Philippe, thanks for addressing this. However, once we enter `if (substType)`, there is a comment which indicates that by the time all clauses of this `if` are processed, `replacedCtxt` should never be null, so I still believe that it being null should be a fatal error. What is the reason to make it only a warning?

```cpp
   if (substType) {
      // Make sure it got replaced from this template
      const clang::ClassTemplateDecl *replacedCtxt = 0;
   ...   
  } else if (...) {
```

Also, this funtion is 250 lines long. It would improve readability a lot if it was broken into smaller pieces.

Reply:
> What is the reason to make it only a warning?

Because I believe that the 'redundant' test 4606 is likely to still give the correct result even if replacedCtxt is a nullptr and even if those two fails, the final code might still do something sensible.